google/resource_compute_instance.go

```diff
@@ -1198,6 +1198,9 @@ func expandInstanceGuestAccelerators(d TerraformResourceData, config *Config) ([
 	guestAccelerators := make([]*computeBeta.AcceleratorConfig, len(accels))
```

Comment:
The issue is that you create empty entries here. Even if you use `continue` below, the empty entries are still added to the list.

Instead, change this line for: `guestAccelerators := make([]*computeBeta.AcceleratorConfig, 0, len(accels))`

And change the line below starting with `guestAccelerators[i] = ...` for `guestAccelerators = append(guestAccelerators, ...)`

Reply:
I just attempted this and the test still fails. My theory now is that the [resourceComputeInstanceRead](https://github.com/jacobstr/terraform-provider-google/blob/bedec8be4bfbdd380dc45bab024ad2e84f837e10/google/resource_compute_instance.go#L723) at the end of [resourceComputeInstanceCreate](https://github.com/jacobstr/terraform-provider-google/blob/bedec8be4bfbdd380dc45bab024ad2e84f837e10/google/resource_compute_instance.go#L856) is what is persisted to terraform's state. 

When the Plan is refreshed it sees `{guest_accelerators: []}` but the current context is requesting `{guest_accelerators: [{count:0, type: "nvidia-tesla-p80"}]}`.

The right way to do this might be to drop/modify the `guest_accelerator` from the `schema.ResourceData` instance as it's being read or immediately afterwards when the count is 0. I'll have to poke if there's an appropriate lifecycle hook (e.g.  afterSchemaResourceDataRead) where this could be implemented.

I'm still puzzled why a similar behavior wasn't observed with the instance template.

Reply:
The current state depends whether the `-refresh` flag is true or false.

When you see a diff like:
```sh
guest_accelerator.#:                    "0" => "1" (forces new resource)
guest_accelerator.0.count:          "" => "0" (forces new resource)
guest_accelerator.0.type:            "" => "nvidia-tesla-k80" (forces new resource)
```

The left hand side is the current state. By default, when you run `terraform plan` or `terraform apply`, the `-refresh=true`. This means it calls the Read function to refresh the current state. If you set `-refresh=false`, then, the current state will be equal to whatever is stored in your state file.

The right hand side (after =>) is always equal to what you have in your Terraform config file (.tf file).

In your case, the config has one `guest_accelerator` entry with `count = 0` and `type = nvidia-tesla-k80`. However, the current state is empty.

You can use the new [customdiff](https://godoc.org/github.com/hashicorp/terraform/helper/customdiff) feature to suppress the diff in that case. I added this new helper to our codebase yesterday and the PR hasn't been merged yet: #945.

Reply:
Thanks @rosbo. Taking a stab with CustomizeDiffFunc.